psdisk/psdisk.c

```diff
@@ -12,156 +12,570 @@
+	}
 
-	/* TODO: call ptable_writePartitions function */
+	/* Verify values from file after writing */
+	i = 0;
+	LIST_FOREACH(pNode, &psdisk_common.list, ptrs) {
+		if (pNode->status == part_save) {
+			if (i >= psdisk_common.pCnt || memcmp(&pNode->pHeader, &pHeaders[i], sizeof(ptable_partition_t)) != 0) {
+				free(pHeaders);
+				return -1;
+			}
+			++i;
+		}
+	}
 
+	free(pHeaders);
 
 	return 0;
 }
 
 
-static int psdisk_parsePartitionOpt(int optind, char *argv[])
+static int psdisk_createPartsTable(void)
 {
 	int i;
-	char *name;
-	ptable_partition_t pHeader;
+	struct plist_node_t *node;
+	ptable_partition_t *pHeaders;
+
+	if ((pHeaders = (ptable_partition_t *)calloc(psdisk_common.pCnt, sizeof(ptable_partition_t))) == NULL) {
+		printf("Cannot allocate memory, err: %s.\n", strerror(errno));
+		return -1;
+	}
+
+	/* Copy partition table to temporary array */
+	i = 0;
+	LIST_FOREACH(node, &psdisk_common.list, ptrs) {
+		if (node->status == part_save)
+			memcpy(&pHeaders[i++], &node->pHeader, sizeof(ptable_partition_t));
+	}
+
+	if (ptable_writePartitions(pHeaders, psdisk_common.pCnt, &psdisk_common.mem) < 0) {
+		printf("Cannot write partition table to file %s.\n", psdisk_common.fileName);
+		free(pHeaders);
+		return -1;
+	}
+
+	free(pHeaders);
+
+	return 0;
+}
 
-	const int MAX_PARTITION_ATTR = 4;
 
-	for (i = 0; i < MAX_PARTITION_ATTR; ++i) {
-		if (argv[optind + i] == NULL) {
-			LOG_ERROR("Invalid number of arguments.");
+static int psdisk_createImg(void)
+{
+	if (psdisk_createPartsTable() < 0)
+		return -1;
+
+	if (psdisk_verifyPartsTable() < 0)
+		return -1;
+
+	psdisk_showPartsTable();
+
+	return 0;
+}
+
+
+static int psdisk_updatePartsList(ptable_partition_t *pHeaders, uint32_t pCnt)
+{
+	int i, action, namesEq;
+	struct plist_node_t *pNode, *newNode, *tempNode;
+
+	for (i = 0; i < pCnt; ++i) {
+		/* Default action: partition from file differs from partition defines by user. */
+		action = part_save;
+
+		LIST_FOREACH(pNode, &psdisk_common.list, ptrs) {
+			namesEq = (strcmp((const char *)&pNode->pHeader.name, (const char *)&pHeaders[i].name) == 0);
+			/* Remove partition declare by user */
+			if(namesEq && pNode->status == part_remove) {
+				action = part_remove;
+				break;
+			}
+			/* Partition from file is updated by new one defines by user. */
+			else if(namesEq && pNode->status == part_save) {
+				action = -1;
+				break;
+			}
+		}
+
+		switch (action) {
+			case part_save:
+				if ((newNode = (struct plist_node_t *)malloc(sizeof(struct plist_node_t))) == NULL) {
+					printf("Cannot allocate memory, err: %s.\n", strerror(errno));
+					return -1;
+				}
+				memcpy(&newNode->pHeader, &pHeaders[i], sizeof(ptable_partition_t));
+				newNode->status = part_save;
+
+				LIST_INSERT_HEAD(&psdisk_common.list, newNode, ptrs);
+				++psdisk_common.pCnt;
+				break;
+
+			case part_remove:
+				tempNode = LIST_NEXT(pNode, ptrs);
+				LIST_REMOVE(pNode, ptrs);
+				free(pNode);
+				pNode = tempNode;
+				--psdisk_common.pCnt;
```

Comment:
Can we achieve negative pCnt this way?

Reply:
Partition with status `part_remove` is added in `psdisk_parseToRm`. At the same time, `pCnt` is increased. So there is no chance to achieve negative `pCnt`.